openedx/openedx-platform
# Improved metadata fetching and new SAML/Shibboleth tests

**Changes in this PR**:
- New integration tests for the shibboleth functionality.
- SAML provider metadata will be automatically fetched when a new provider is added via the admin interface. This way, for example, if you add a new provider, its metadata will be retrieved so users can authenticate with it immediately.
- SAML Provider metadata will be automatically fetched daily (scheduled task via celery beat).

**Testing on devstack**:
1. Go to http://localhost:8000/admin/third_party_auth/samlproviderconfig/ and configure at least one provider. It doesn't have to be a real provider, but adding a fake one and a real one like TestShib would be best.
2. You should have a working celery setup. If you don't, install rabbitmq (`sudo apt-get install rabbitmq-server`, `sudo rabbitmqctl add_user celery celery`, `sudo rabbitmqctl set_permissions -p / celery ".*" ".*" ".*"`)
3. Edit lms/envs/private.py to set `CELERY_ALWAYS_EAGER = False`.
4. Edit ~/lms.env.json: in the `CACHES` section, add a new entry for `configuration`:
   `json
       "configuration": {
           "BACKEND": "django.core.cache.backends.memcached.MemcachedCache",
           "KEY_FUNCTION": "util.memcache.safe_key",
           "KEY_PREFIX": "configuration",
           "LOCATION": [
               "localhost:11211"
           ]
       },
   `
5. Also in ~/lms.env.json, add a new top-level setting `"THIRD_PARTY_AUTH_SAML_FETCH_PERIOD_HOURS": 0.01,`
6. Start a celery worker `paver celery --settings=devstack`. You should see it run the `fetch_saml_metadata` task and log the output to the console (`Scheduler: Sending due task refresh-saml-metadata` and `Fetching https://www.testshib.org/metadata/testshib-providers.xml`, `→ Updated existing SAMLProviderData. Nothing has changed.` etc.). Then, 36 seconds later, you should see the task repeat.
7. Delete the `THIRD_PARTY_AUTH_SAML_FETCH_PERIOD_HOURS` setting, to restore the default value of 24 hours, and restart the celery worker.
8. Run the LMS while the celery worker is running. Make a change to the SAML provider configuration, ideally enabling or disabling one of the providers.  You should see that soon after you make a change, the celery task will run and update the providers' metadata.
9. The new integration tests are tested by our friend Jenkins.

**Misc**
No JIRA story, sandbox, screenshots, or dependencies.

**Notes**:
On a production deployment, celery may need to be configured to run a `beat` process. If using paver to launch the workers, that is achieved with this PR, but if the workers are started some other way, the `--beat` parameter will need to be added if SAML fetching or other scheduled tasks are desired.

**Reviewers**:
@smarnach and Cale
